collectors/CMPCollector.js

```diff
@@ -24,70 +45,292 @@ class CMPCollector extends BaseCollector {
+                        const msg = JSON.parse(payload);
+                        // this.log(`autoconsent message from ${executionContextId}: ${JSON.stringify(msg)}`);
+                        await this.handleMessage(msg, executionContextId);
+                    } catch (e) {
+                        this.log(`Could not handle autoconsent message ${payload}`);
+                    }
+                }
+            });
+            await this._cdpClient.send('Runtime.addBinding', {
+                name: 'autoconsentStandaloneSendMessage',
+                executionContextName: worldName,
+            });
+        }
+    }
+
+    /**
+     * Implements autoconsent messaging protocol
+     *
+     * @param {import('@duckduckgo/autoconsent/lib/messages').ContentScriptMessage} msg
+     * @param {any} executionContextId
+     * @returns {Promise<void>}
+     */
+    async handleMessage(msg, executionContextId) {
+        this.receivedMsgs.push(msg);
+        switch (msg.type) {
+        case 'optInResult':
+        case 'optOutResult': {
+            if (msg.scheduleSelfTest) {
+                this.selfTestFrame = executionContextId;
+            }
+            break;
+        }
+        case 'autoconsentDone': {
+            if (this.selfTestFrame) {
+                await this._cdpClient.send('Runtime.evaluate', {
+                    expression: `autoconsentStandaloneReceiveMessage({ type: "selfTest" })`,
+                    allowUnsafeEvalBlockedByCSP: true,
+                    contextId: this.selfTestFrame,
                 });
-            }, () => this.log('Unable to get pages'));
+            }
+            break;
+        }
+        case 'eval': {
+            let evalResult = false;
+            const result = await this._cdpClient.send('Runtime.evaluate', {
+                expression: msg.code,
+                returnByValue: true,
+                allowUnsafeEvalBlockedByCSP: true,
+                contextId: this.isolated2pageworld.get(executionContextId), // this must be done in page world
+            });
+            if (!result.exceptionDetails) {
+                evalResult = Boolean(result.result.value);
+            }
+
+            await this._cdpClient.send('Runtime.evaluate', {
+                expression: `autoconsentStandaloneReceiveMessage({ id: "${msg.id}", type: "evalResp", result: ${JSON.stringify(evalResult)} })`,
+                allowUnsafeEvalBlockedByCSP: true,
+                contextId: executionContextId,
+            });
+            break;
+        }
+        case 'autoconsentError': {
+            this.log(`autoconsent error: ${msg.details}`);
+            break;
+        }
+        default:
+        }
+    }
+
+    /**
+     * @param {Partial<import('@duckduckgo/autoconsent/lib/messages').ContentScriptMessage>} msg
+     * @returns {Promise<import('@duckduckgo/autoconsent/lib/messages').ContentScriptMessage>}
+     */
+    async waitForMessage(msg, maxTimes = 20, interval = 100) {
```

Comment:
I believe this may be too little waiting time - when running a coverage crawl it looks like a load of sites were not detected, probably because we returned an empty result before the CMP actually got a chance to render.

Reply:
Before calling `collector.getData()`, crawler [waits](https://github.com/duckduckgo/tracker-radar-collector/blob/new-autoconsent/crawler.js#L202) for the `onload` event (up to 30s), and then [guaranteed](https://github.com/duckduckgo/tracker-radar-collector/blob/new-autoconsent/crawler.js#L220) 2.5s after that. I figured that should have been enough. Besides, `waitForMessage()` timeout is for a single message, and we call it multiple times.
That said, it's easy to test this hypothesis. I will try to increase the timeout for the very first message and run the coverage crawl again.

Reply:
We tested this and increasing timeouts doesn't affect the results.